Cannot configure endpoint for AWS Secrets Manager or Parameter Store

**Describe the bug**
Cannot configure endpoint for AWS Secrets Manager or Parameter Store integration.  The application fails to start with the following exception

```
org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'awsSecretsManagerEnvironmentRepository' defined in class path resource [org/springframework/cloud/config/server/config/AwsSecretsManagerRepositoryConfiguration.class]: Bean instantiation via factory method failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [org.springframework.cloud.config.server.environment.AwsSecretsManagerEnvironmentRepository]: Factory method 'awsSecretsManagerEnvironmentRepository' threw exception; nested exception is java.lang.IllegalStateException: Only one of Region or EndpointConfiguration may be set.
	at org.springframework.beans.factory.support.ConstructorResolver.instantiate(ConstructorResolver.java:658) ~[spring-beans-5.3.13.jar:5.3.13]
	at org.springframework.beans.factory.support.ConstructorResolver.instantiateUsingFactoryMethod(ConstructorResolver.java:638) ~[spring-beans-5.3.13.jar:5.3.13]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.instantiateUsingFactoryMethod(AbstractAutowireCapableBeanFactory.java:1352) ~[spring-beans-5.3.13.jar:5.3.13]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBeanInstance(AbstractAutowireCapableBeanFactory.java:1195) ~[spring-beans-5.3.13.jar:5.3.13]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:582) ~[spring-beans-5.3.13.jar:5.3.13]
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:542) ~[spring-beans-5.3.13.jar:5.3.13]
	at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:335) ~[spring-beans-5.3.13.jar:5.3.13]
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:234) ~[spring-beans-5.3.13.jar:5.3.13]
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:333) ~[spring-beans-5.3.13.jar:5.3.13]
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:208) ~[spring-beans-5.3.13.jar:5.3.13]
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.preInstantiateSingletons(DefaultListableBeanFactory.java:944) ~[spring-beans-5.3.13.jar:5.3.13]
	at org.springframework.context.support.AbstractApplicationContext.finishBeanFactoryInitialization(AbstractApplicationContext.java:918) ~[spring-context-5.3.13.jar:5.3.13]
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:583) ~[spring-context-5.3.13.jar:5.3.13]
	at org.springframework.boot.web.servlet.context.ServletWebServerApplicationContext.refresh(ServletWebServerApplicationContext.java:145) ~[spring-boot-2.6.1.jar:2.6.1]
	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:730) ~[spring-boot-2.6.1.jar:2.6.1]
	at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:412) ~[spring-boot-2.6.1.jar:2.6.1]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:302) ~[spring-boot-2.6.1.jar:2.6.1]
	at org.springframework.boot.builder.SpringApplicationBuilder.run(SpringApplicationBuilder.java:164) ~[spring-boot-2.6.1.jar:2.6.1]
	at org.springframework.cloud.config.server.ConfigServerApplication.main(ConfigServerApplication.java:35) ~[spring-cloud-config-server-3.1.0.jar:3.1.0]
Caused by: org.springframework.beans.BeanInstantiationException: Failed to instantiate [org.springframework.cloud.config.server.environment.AwsSecretsManagerEnvironmentRepository]: Factory method 'awsSecretsManagerEnvironmentRepository' threw exception; nested exception is java.lang.IllegalStateException: Only one of Region or EndpointConfiguration may be set.
	at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:185) ~[spring-beans-5.3.13.jar:5.3.13]
	at org.springframework.beans.factory.support.ConstructorResolver.instantiate(ConstructorResolver.java:653) ~[spring-beans-5.3.13.jar:5.3.13]
	... 18 common frames omitted
Caused by: java.lang.IllegalStateException: Only one of Region or EndpointConfiguration may be set.
	at com.amazonaws.client.builder.AwsClientBuilder.setRegion(AwsClientBuilder.java:450) ~[aws-java-sdk-core-1.11.935.jar:na]
	at com.amazonaws.client.builder.AwsClientBuilder.configureMutableProperties(AwsClientBuilder.java:424) ~[aws-java-sdk-core-1.11.935.jar:na]
	at com.amazonaws.client.builder.AwsSyncClientBuilder.build(AwsSyncClientBuilder.java:46) ~[aws-java-sdk-core-1.11.935.jar:na]
	at org.springframework.cloud.config.server.environment.AwsSecretsManagerEnvironmentRepositoryFactory.build(AwsSecretsManagerEnvironmentRepositoryFactory.java:56) ~[spring-cloud-config-server-3.1.0.jar:3.1.0]
	at org.springframework.cloud.config.server.config.AwsSecretsManagerRepositoryConfiguration.awsSecretsManagerEnvironmentRepository(EnvironmentRepositoryConfiguration.java:438) ~[spring-cloud-config-server-3.1.0.jar:3.1.0]
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:na]
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[na:na]
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:na]
	at java.base/java.lang.reflect.Method.invoke(Method.java:566) ~[na:na]
	at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:154) ~[spring-beans-5.3.13.jar:5.3.13]
	... 19 common frames omitted

```

The problem is `org.springframework.cloud.config.server.environment.AwsSecretsManagerEnvironmentRepositoryFactory` and `org.springframework.cloud.config.server.environment.AwsParameterStoreEnvironmentRepositoryFactory` do not configure the region in the AwsClientBuilder.EndpointConfiguration, as is done in `org.springframework.cloud.config.server.environment.AwsS3EnvironmentRepositoryFactory`

I have submitted PR #2023 to fix it

**Sample**
```yaml
spring:
  profiles:
    active: awssecretsmanager
  cloud:
    config:
      server:
        aws-secretsmanager:
          endpoint: localhost.localstack.cloud:4566
          region: us-east-1

```

```yaml
spring:
  profiles:
    active: awsparamstore
  cloud:
    config:
      server:
        awsparamstore:
          endpoint: localhost.localstack.cloud:4566
          region: us-east-1

```

--- a/spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/environment/AwsClientBuilderConfigurer.java
+++ b/spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/environment/AwsClientBuilderConfigurer.java
@@ -1,0 +1,40 @@
+/*
+ * Copyright 2018-2020 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.config.server.environment;
+
+import com.amazonaws.client.builder.AwsClientBuilder;
+import com.amazonaws.client.builder.AwsSyncClientBuilder;
+
+import org.springframework.util.StringUtils;
+
+abstract class AwsClientBuilderConfigurer {
+	private AwsClientBuilderConfigurer() {
+	}
+
+	static void configureClientBuilder(AwsSyncClientBuilder<?, ?> clientBuilder, String region, String endpoint) {
+		if (StringUtils.hasText(region)) {
+			if (StringUtils.hasText(endpoint)) {
+				AwsClientBuilder.EndpointConfiguration endpointConfiguration = new AwsClientBuilder.EndpointConfiguration(
+					endpoint, region);
+				clientBuilder.withEndpointConfiguration(endpointConfiguration);
+			}
+			else {
+				clientBuilder.withRegion(region);
+			}
+		}
+	}
+}

--- a/spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/environment/AwsParameterStoreEnvironmentRepositoryFactory.java
+++ b/spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/environment/AwsParameterStoreEnvironmentRepositoryFactory.java
@@ -16,13 +16,12 @@
 
 package org.springframework.cloud.config.server.environment;
 
-import com.amazonaws.client.builder.AwsClientBuilder;
-import com.amazonaws.regions.Regions;
 import com.amazonaws.services.simplesystemsmanagement.AWSSimpleSystemsManagement;
 import com.amazonaws.services.simplesystemsmanagement.AWSSimpleSystemsManagementClientBuilder;
 
 import org.springframework.cloud.config.server.config.ConfigServerProperties;
-import org.springframework.util.StringUtils;
+
+import static org.springframework.cloud.config.server.environment.AwsClientBuilderConfigurer.configureClientBuilder;
 
 /**
  * @author Iulian Antohe
@@ -40,22 +39,7 @@ public class AwsParameterStoreEnvironmentRepositoryFactory implements
 	public AwsParameterStoreEnvironmentRepository build(AwsParameterStoreEnvironmentProperties environmentProperties) {
 		AWSSimpleSystemsManagementClientBuilder clientBuilder = AWSSimpleSystemsManagementClientBuilder.standard();
 
-		String region = environmentProperties.getRegion();
-
-		if (StringUtils.hasLength(region)) {
-			Regions awsRegion = Regions.fromName(region);
-
-			clientBuilder.withRegion(awsRegion);
-
-			String endpoint = environmentProperties.getEndpoint();
-
-			if (StringUtils.hasLength(endpoint)) {
-				AwsClientBuilder.EndpointConfiguration endpointConfiguration = new AwsClientBuilder.EndpointConfiguration(
-						endpoint, awsRegion.getName());
-
-				clientBuilder.withEndpointConfiguration(endpointConfiguration);
-			}
-		}
+		configureClientBuilder(clientBuilder, environmentProperties.getRegion(), environmentProperties.getEndpoint());
 
 		AWSSimpleSystemsManagement client = clientBuilder.build();
 

--- a/spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/environment/AwsS3EnvironmentRepositoryFactory.java
+++ b/spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/environment/AwsS3EnvironmentRepositoryFactory.java
@@ -21,6 +21,8 @@ import com.amazonaws.services.s3.AmazonS3ClientBuilder;
 
 import org.springframework.cloud.config.server.config.ConfigServerProperties;
 
+import static org.springframework.cloud.config.server.environment.AwsClientBuilderConfigurer.configureClientBuilder;
+
 public class AwsS3EnvironmentRepositoryFactory
 		implements EnvironmentRepositoryFactory<AwsS3EnvironmentRepository, AwsS3EnvironmentProperties> {
 
@@ -33,16 +35,7 @@ public class AwsS3EnvironmentRepositoryFactory
 	@Override
 	public AwsS3EnvironmentRepository build(AwsS3EnvironmentProperties environmentProperties) {
 		final AmazonS3ClientBuilder clientBuilder = AmazonS3ClientBuilder.standard();
-		if (environmentProperties.getRegion() != null) {
-			if (environmentProperties.getEndpoint() != null) {
-				AmazonS3ClientBuilder.EndpointConfiguration endpointConfiguration = new AmazonS3ClientBuilder.EndpointConfiguration(
-						environmentProperties.getEndpoint(), environmentProperties.getRegion());
-				clientBuilder.withEndpointConfiguration(endpointConfiguration);
-			}
-			else {
-				clientBuilder.withRegion(environmentProperties.getRegion());
-			}
-		}
+		configureClientBuilder(clientBuilder, environmentProperties.getRegion(), environmentProperties.getEndpoint());
 		final AmazonS3 client = clientBuilder.build();
 
 		AwsS3EnvironmentRepository repository = new AwsS3EnvironmentRepository(client,

--- a/spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/environment/AwsSecretsManagerEnvironmentRepositoryFactory.java
+++ b/spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/environment/AwsSecretsManagerEnvironmentRepositoryFactory.java
@@ -16,13 +16,12 @@
 
 package org.springframework.cloud.config.server.environment;
 
-import com.amazonaws.client.builder.AwsClientBuilder;
-import com.amazonaws.regions.Regions;
 import com.amazonaws.services.secretsmanager.AWSSecretsManager;
 import com.amazonaws.services.secretsmanager.AWSSecretsManagerClientBuilder;
 
 import org.springframework.cloud.config.server.config.ConfigServerProperties;
-import org.springframework.util.StringUtils;
+
+import static org.springframework.cloud.config.server.environment.AwsClientBuilderConfigurer.configureClientBuilder;
 
 /**
  * @author Tejas Pandilwar
@@ -39,19 +38,8 @@ public class AwsSecretsManagerEnvironmentRepositoryFactory implements
 	@Override
 	public AwsSecretsManagerEnvironmentRepository build(AwsSecretsManagerEnvironmentProperties environmentProperties) {
 		AWSSecretsManagerClientBuilder clientBuilder = AWSSecretsManagerClientBuilder.standard();
-		String region = environmentProperties.getRegion();
 
-		if (!StringUtils.isEmpty(region)) {
-			Regions awsRegion = Regions.fromName(region);
-			clientBuilder.withRegion(awsRegion);
-
-			String endpoint = environmentProperties.getEndpoint();
-			if (!StringUtils.isEmpty(endpoint)) {
-				AwsClientBuilder.EndpointConfiguration endpointConfiguration = new AwsClientBuilder.EndpointConfiguration(
-						endpoint, awsRegion.getName());
-				clientBuilder.withEndpointConfiguration(endpointConfiguration);
-			}
-		}
+		configureClientBuilder(clientBuilder, environmentProperties.getRegion(), environmentProperties.getEndpoint());
 
 		AWSSecretsManager client = clientBuilder.build();
 		return new AwsSecretsManagerEnvironmentRepository(client, configServerProperties, environmentProperties);

--- a/spring-cloud-config-server/src/test/java/org/springframework/cloud/config/server/environment/AwsParameterStoreEnvironmentRepositoryTests.java
+++ b/spring-cloud-config-server/src/test/java/org/springframework/cloud/config/server/environment/AwsParameterStoreEnvironmentRepositoryTests.java
@@ -695,6 +695,25 @@ public class AwsParameterStoreEnvironmentRepositoryTests {
 		assertThat(result).usingRecursiveComparison().withStrictTypeChecking().isEqualTo(expected);
 	}
 
+	@Test
+	public void factoryCustomizableWithRegion() {
+		AwsParameterStoreEnvironmentRepositoryFactory factory = new AwsParameterStoreEnvironmentRepositoryFactory(new ConfigServerProperties());
+		AwsParameterStoreEnvironmentProperties properties = new AwsParameterStoreEnvironmentProperties();
+		properties.setRegion("us-east-1");
+		AwsParameterStoreEnvironmentRepository repository = factory.build(properties);
+		assertThat(repository).isNotNull();
+	}
+
+	@Test
+	public void factoryCustomizableWithRegionAndEndpoint() {
+		AwsParameterStoreEnvironmentRepositoryFactory factory = new AwsParameterStoreEnvironmentRepositoryFactory(new ConfigServerProperties());
+		AwsParameterStoreEnvironmentProperties properties = new AwsParameterStoreEnvironmentProperties();
+		properties.setRegion("us-east-1");
+		properties.setEndpoint("https://myawsendpoint/");
+		AwsParameterStoreEnvironmentRepository repository = factory.build(properties);
+		assertThat(repository).isNotNull();
+	}
+
 	private void setupAwsSsmClientMocks(Environment environment) {
 		setupAwsSsmClientMocks(environment, false, false);
 	}

--- a/spring-cloud-config-server/src/test/java/org/springframework/cloud/config/server/environment/AwsSecretsManagerEnvironmentRepositoryTests.java
+++ b/spring-cloud-config-server/src/test/java/org/springframework/cloud/config/server/environment/AwsSecretsManagerEnvironmentRepositoryTests.java
@@ -715,6 +715,25 @@ public class AwsSecretsManagerEnvironmentRepositoryTests {
 		assertThat(resultEnv).usingRecursiveComparison().withStrictTypeChecking().isEqualTo(expectedEnv);
 	}
 
+	@Test
+	public void factoryCustomizableWithRegion() {
+		AwsSecretsManagerEnvironmentRepositoryFactory factory = new AwsSecretsManagerEnvironmentRepositoryFactory(new ConfigServerProperties());
+		AwsSecretsManagerEnvironmentProperties properties = new AwsSecretsManagerEnvironmentProperties();
+		properties.setRegion("us-east-1");
+		AwsSecretsManagerEnvironmentRepository repository = factory.build(properties);
+		assertThat(repository).isNotNull();
+	}
+
+	@Test
+	public void factoryCustomizableWithRegionAndEndpoint() {
+		AwsSecretsManagerEnvironmentRepositoryFactory factory = new AwsSecretsManagerEnvironmentRepositoryFactory(new ConfigServerProperties());
+		AwsSecretsManagerEnvironmentProperties properties = new AwsSecretsManagerEnvironmentProperties();
+		properties.setRegion("us-east-1");
+		properties.setEndpoint("https://myawsendpoint/");
+		AwsSecretsManagerEnvironmentRepository repository = factory.build(properties);
+		assertThat(repository).isNotNull();
+	}
+
 	private void setupAwsSmClientMocks(Environment environment) {
 		for (PropertySource ps : environment.getPropertySources()) {
 			String path = StringUtils.delete(ps.getName(), environmentProperties.getOrigin());

